wallet-new/src/Cardano/Wallet/API/Indices.hs

```diff
@@ -39,6 +41,12 @@ type family IndicesOf a :: [*] where
     IndicesOf Transaction = TransactionIxs
     IndicesOf Account     = AccountIxs
 
+-- | This type family imposes a bidirectional mapping between symbols and
+-- their index types. This allows you to go from a symbol to the type:
+type family StringToIndex sym = r | r -> sym  where
+    StringToIndex "balance" = Core.Coin
+    StringToIndex "wallet_id" = WalletId
```

Comment:
@parsonsmatt Unfortunately I'm not sure this is gonna work out without indexing the type family via `r` as well. If we ever want to have overloaded params (something we can avoid but would still be nice), you will have clashes. For example, in this PR Konstantin is now introducing the `id` filter parameter to filter on `Transaction`s:

https://github.com/input-output-hk/cardano-sl/pull/2602/files#diff-8818902a6e1eac32f98a94cdd65d7048R10

One might argue we should fully qualify our query params (and that would be a perfectly reasonable answer), but it would be nice if we could simply say something like this:

```
GET /api/v1/wallets?id=Blabla
GET /api/v1/transactions?id=blabla
```

In the RESTful sense, as we are dealing with well-defined resources (in this example `Wallet` and `Transaction`) there is no ambiguity (at least imho) on which type of `id` we are referring to. But in this new brave world `StringToIndex` won't work, but this will:

```haskell
type family StringToIndex sym r = where
  StringToIndex "id" Wallet = WalletId
  StringToIndex "id" Transaction = TransactionId
```

I'm ready to be told it's a terrible idea, and you might be right 😁 What I think is most important, though, is to choose a schema and stick with it. If we are opting for "fully qualified" query params is in `resource` + `field_identifier` better have this added to the docs for all the developers to abide 😛

Reply:
Or, we make the tying explicit at the API definition level:

```haskell
data (sym :: Symbol) ?= (a :: *)

data FilterBy params resource

type Api
  = "wallets"
    :> FilterBy '[ "id" ?= WalletId, "balance" ?= Core.Coin ] Wallet
    :> Get '[JSON] [Wallet]
  :<|> "transactions"
    :> FilterBy '[ "id ?= TransactionId ] Transaction
    :> Get '[JSON] [Transaction] 
```

Then the relevant instances have both the `Symbol` and the the type available.

Reply:
Fully qualified API names sounds awful, IMO, so you're right, this isn't a good design, except as a stopgap to get this implemented now.

Reply:
> Or, we make the tying explicit at the API definition level:

Sweet combinator! The only little problem might be we are now giving the developer the degree of freedom or reassociating the symbol `id` to `WalletId` or any other index, but I think this is totally fine as the type family we have on mapping symbols to `ixset` indices should catch the bug. I think I like the extra verbosity of specifying the type in the RHS of `?=` as it's the kind of good verbosity imho!